pkg/cmd/cli/cmd/set/triggers.go

```diff
@@ -0,0 +1,740 @@
+				return fmt.Errorf("you must specify --containers when setting --from-image")
+			}
+			if !allNames.HasAll(trigger.Names...) {
+				return fmt.Errorf(
+					"not all container names exist: %s (accepts: %s)",
+					strings.Join(sets.NewString(trigger.Names...).Difference(allNames).List(), ", "),
+					strings.Join(allNames.List(), ", "),
+				)
+			}
+			triggers = append(triggers, deployapi.DeploymentTriggerPolicy{
+				Type: deployapi.DeploymentTriggerOnImageChange,
+				ImageChangeParams: &deployapi.DeploymentTriggerImageChangeParams{
+					Automatic: trigger.Auto,
+					From: kapi.ObjectReference{
+						Kind: "ImageStreamTag",
+						Name: trigger.From,
+					},
+					ContainerNames: trigger.Names,
+				},
+			})
+		}
+		c.Spec.Triggers = mergeDeployTriggers(existingTriggers, triggers)
+		return nil
+
+	case *buildapi.BuildConfig:
+		var triggers []buildapi.BuildTriggerPolicy
+		if t.ConfigChange {
+			triggers = append(triggers, buildapi.BuildTriggerPolicy{Type: buildapi.ConfigChangeBuildTriggerType})
+		}
+		for _, trigger := range t.WebHooks {
+			triggers = append(triggers, buildapi.BuildTriggerPolicy{
+				Type: buildapi.GenericWebHookBuildTriggerType,
+				GenericWebHook: &buildapi.WebHookTrigger{
+					Secret: trigger,
+				},
+			})
+		}
+		for _, trigger := range t.GitHubWebHooks {
+			triggers = append(triggers, buildapi.BuildTriggerPolicy{
+				Type: buildapi.GitHubWebHookBuildTriggerType,
+				GitHubWebHook: &buildapi.WebHookTrigger{
+					Secret: trigger,
+				},
+			})
+		}
+
+		// add new triggers, filter out any old triggers that match (if moving from automatic to manual),
+		// and then merge the old triggers and the new triggers to preserve fields like lastTriggeredImageID
+		existingTriggers := c.Spec.Triggers
+		strategyTrigger := strategyTrigger(c)
+		for _, trigger := range t.ImageChange {
+			change := &buildapi.ImageChangeTrigger{
+				From: &kapi.ObjectReference{
+					Kind:      "ImageStreamTag",
+					Name:      trigger.From,
+					Namespace: trigger.Namespace,
+				},
+			}
+
+			// use the canonical ImageChangeTrigger with nil From
+			strategyTrigger.Auto = trigger.Auto
+			if reflect.DeepEqual(strategyTrigger, &trigger) {
+				change.From = nil
+			}
+
+			// if this trigger is not automatic, then we need to remove it from the list of triggers
+			if !trigger.Auto {
+				existingTriggers = filterBuildImageTriggers(existingTriggers, trigger, strategyTrigger)
+				continue
+			}
+
+			triggers = append(triggers, buildapi.BuildTriggerPolicy{
+				Type:        buildapi.ImageChangeBuildTriggerType,
+				ImageChange: change,
+			})
+		}
+		c.Spec.Triggers = mergeBuildTriggers(existingTriggers, triggers)
+		return nil
+
+	default:
+		return fmt.Errorf("the object is not a deployment config or build config")
+	}
+}
+
+func triggerMatchesBuildImageChange(trigger ImageChangeTrigger, strategyTrigger *ImageChangeTrigger, imageChange *buildapi.ImageChangeTrigger) bool {
+	if imageChange == nil {
```

Comment:
Logic was actually wrong, strategy trigger should be able to be nil.  Updated in review 2